### PR TITLE
Add basic nameplate service

### PR DESCRIPTION
**Pull request by Pilzinsel64**
grafik

Keep in mind that this is just a basic (but futureorientate) implementation, re-implemented inspired by Pilz.Dalamud. The current implementation uses the `SetPlayerNameplateDetourDelegate` event and provides just a basic way to edit the existing elements of a player nameplate.

It's not possible yet to directly get/edit/add/remove nodes. This can be done in a future PR by a future implementation of the Nameplate service wich uses another Event (more low-level wich then access AktResNode etc.). But I had not the time to get involved with this Event. Most plugins uses this one and that's enough for now imao.

I know the other event is more low-level that gives you more options. But I have no expierience with it and time to investigate and it's not really required for this basic PR in my opinion.

Also, the manual allocation has it's reason. `MemoryHelper.WriteString` didn't work. The parameters like `titlePtr` are addresses to raw byte arrays, so `Utf8String` isn't working too.

### CODE REVIEW

File: Dalamud/Game/Gui/Nameplates/EventArgs/NameplateUpdateEventArgs.cs
Comment (MidoriKami) [inline Dalamud/Game/Gui/Nameplates/EventArgs/NameplateUpdateEventArgs.cs:27]: Per <https://github.com/goatcorp/Dalamud/pull/1808> It'll probably be best to create an interface for this type.

File: Dalamud/Game/Gui/Nameplates/Model/NameplateInfo.cs
Comment (MidoriKami) [inline Dalamud/Game/Gui/Nameplates/Model/NameplateInfo.cs:46]: Per <https://github.com/goatcorp/Dalamud/pull/1808> It'll probably be best to create an interface for this type.

File: Dalamud/Game/Gui/Nameplates/Model/NameplateObject.cs
Comment (MidoriKami) [inline Dalamud/Game/Gui/Nameplates/Model/NameplateObject.cs:13]: Per <https://github.com/goatcorp/Dalamud/pull/1808> It'll probably be best to create an interface for this type.

File: Dalamud/Game/Gui/Nameplates/Model/StatusIcons.cs
Comment (MidoriKami) [inline Dalamud/Game/Gui/Nameplates/Model/StatusIcons.cs:91]: Does this mimic the datasheet for `OnlineStatus`?

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (MidoriKami) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:43]: This is likely to explode when the player logs out and logs back in.

The Addon is re-created when you login.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (MidoriKami) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:126]: Is it necessary to parse this data every time? This is a lot of SeString parsing that is happening when it might not even be required. Might be better for NameplateInfo to store a address internally and offer this parsing as a get-only property.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (MidoriKami) [reply Dalamud/Game/Gui/Nameplates/NameplateGui.cs:43]: Possible fix includes using AddonLifcycle to listen for the Setup and Finalize events for NamePlate, and use that to manage the pointer.

**Comment by MidoriKami** [comment]
@Infiziert90 noticed that you are using `IntPtr` everywhere, should use `nint` instead.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (Haselnussbomber) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:160]: Please do not write to the pointers.
The data comes from StringArrayData and you don't know how much space the game has allocated for these strings.
Instead, you should use `Unsafe.AsPointer` to get pointers of your encoded byte arrays and pass them to the Original function. There is no need to free them, since the data is on the stack.
The only thing you probably have to do is to make sure they have a null terminator at the end, so concat the arrays with `new byte[] { 0 }`.

File: Dalamud/Game/Gui/Nameplates/EventArgs/NameplateUpdateEventArgs.cs
Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/EventArgs/NameplateUpdateEventArgs.cs]: Done!

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/NameplateGui.cs]: True. Moved that to `NameplateNode.Text`. It will be parsed when the getter is called the first time.

**Comment by Pilzinsel64** [comment]
> @Infiziert90 noticed that you are using `IntPtr` everywhere, should use `nint` instead.

Changed everything to nint now.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/NameplateGui.cs]: Done!

File: Dalamud/Game/Gui/Nameplates/Model/NameplateInfo.cs
Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/Model/NameplateInfo.cs]: Done!

File: Dalamud/Game/Gui/Nameplates/EventArgs/NameplateUpdateEventArgs.cs
Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/EventArgs/NameplateUpdateEventArgs.cs]: Well, I removed that event args. Never seen them as interface and even don't need them anymore now.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (Pilzinsel64) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:97]: Seems like this doesn't work anymore. Any idea?

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (Haselnussbomber) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:155]: Btw, the reason I told you to use Utf8String is because it will add a null terminator after the last byte of the string.
The way you did it now doesn’t have that, so it doesn’t know where the string ends.
You can just push Utf8Strings to the list and then call Dispose in the ForEach

Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/NameplateGui.cs:156]: Tried that but can't call `Dispose` as it's non-public.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (Haselnussbomber) [reply Dalamud/Game/Gui/Nameplates/NameplateGui.cs:97]: Should still be there. It’s inherited from UIModuleInterface now via the generators. Just build it and maybe restart your IDE.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/NameplateGui.cs:156]: Oh, do you mean `Utfa8String.Dtor(bool)`? Missed that before.

Comment (Haselnussbomber) [reply Dalamud/Game/Gui/Nameplates/NameplateGui.cs:156]: Might have to cast it to IDisposable, like
`(utf8string as IDisposable).Dispose()`

Comment (Haselnussbomber) [reply Dalamud/Game/Gui/Nameplates/NameplateGui.cs:156]: Dispose calls Dtor(false), so that would work too, yep,

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (Haselnussbomber) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:173]: You wrote Dtor(true), this will backfire. You didn’t let the game allocate the Utf8Strings, so it can’t free them. Call Dtor(false), so that it will free the StringPtr inside if needed (I say that because Utf8String has an internal buffer too) and let C#s garbage collector handle the Utf8Strings.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/NameplateGui.cs]: I _think_ I got it now. Thanks! I even could simplify the method `GetNameplateGameObject` a bit more.

File: Dalamud/Game/Gui/Nameplates/Model/INameplateNode.cs
Comment (nebel) [inline Dalamud/Game/Gui/Nameplates/Model/INameplateNode.cs:8]: I wonder if it's better to just call this NameplateUpdateString or NameplateUpdateArgument or something else which doesn't use "node". When I think "node" in terms of the addons like NamePlate I think of the persistent UI nodes which make up the plates.

Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/Model/INameplateNode.cs:8]: I used "Element" before. But then I used "Node" as it the potential to match future expansion of the code.

**Comment by MidoriKami** [comment]
Is there a particular reason or reasons why this is still a draft?

**Comment by Pilzinsel64** [comment]
@MidoriKami I didn't actually tested this via the Tesplugin yet as I had no time this week. :/
Aaand I didn't took a look to the Status Icon Excel Sheet you asked me for to check.

**Comment by MidoriKami** [comment]
Understandable, not trying to rush, it just seemed like it was pretty close to being good to go, we can always make minor adjustments later.

**Comment by Pilzinsel64** [comment]
Alright, ready for review/merge now!
grafik

Keep in mind that this is just a basic (but futureorientate) implementation, re-implemented inspired by Pilz.Dalamud.

It's not possible yet to directly get/edit/add/remove nodes. This can be done by a future implementation of the Nameplate service wich uses another Event (more low-level wich then access AktResNode etc.). But I had not the time to get involved with this Event. Most plugins uses this one and that's enough for now imao.

Also, the manual allocation has it's reason. `MemoryHelper.WriteString` didn't work. The parameters like `titlePtr` are addresses to raw byte arrays, so `Utf8String` isn't working too.

File: Dalamud/Game/Gui/Nameplates/Model/StatusIcons.cs
Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/Model/StatusIcons.cs]: To be checked after merge. Had not the time until feature freeze. But can be deprecated in the future easily, if there is another way to provide that.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (nebel) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:76]: I wonder if it makes sense to cache the result of this in `NameplateObject` or similar. I think finding either the index or the object is going to be necessary for most users as that's how they're likely to either find the nameplate owner. As-is each consumer is likely to repeat this call, so we'll end up looking up the index and scanning the object table multiple times.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (nebel) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:91]: I think it might be a good idea to provide index lookup as a separate function independent from GameObject lookup. If a plugin is doing its own modifications to plates it likely needs to know the index so it can look up the UI Node or find certain data in whatever index it uses internally.

File: Dalamud/Game/Gui/Nameplates/Model/INameplateNode.cs
Comment (nebel) [reply Dalamud/Game/Gui/Nameplates/Model/INameplateNode.cs:8]: I see. I think if actual UI nodes are ever added to this there will likely need to be new field/event/etc. types, so I'm not sure there is really any forwards-compatibility to calling these "Nodes" here. If not "arguments", maybe "elements"? Things like prefix, name and free company are actually concatenated into the final name string and aren't even separate nodes in the final plate.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/NameplateGui.cs]: Yup, makes sence. I removed the generic type and added a property to `INameplateObject`. Also removed `GetNameplateGameObject<T>` from `INameplateGui`.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/NameplateGui.cs:95]: Maybe also makes sense, your right. I added a property to `INameplateObject`. Also splitted `NameplateGui.GetNameplateGameObject<T>` into two methods.

**Comment by kalilistic** [comment]
Will this be merged in with apix? No pressure, but we're talking about falling back to the library if not and want to plan accordingly.

**Comment by goaaats** [comment]
That would be nice, but it'd have to go in this weekend. Could you please re-review if you have time? @Haselnussbomber, @nebel and @MidoriKami - thank you!

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (MidoriKami) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:32]: This should be in a AddressResolver class, see `AddonLifecycleAddressResolver`, and `AddonLifecycle` for an example.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (MidoriKami) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:60]: You need to also Unregister the AddonLifecycle Listeners here.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (MidoriKami) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:116]: I strongly suggest using two different function for Setup and Finalize, it'll make intent more clear and help prevent bugs like what was fixed earlier.

However this is just a style suggestion, feel free to disregard if you prefer.

File: Dalamud/Game/Gui/Nameplates/NameplateGui.cs
Comment (MidoriKami) [inline Dalamud/Game/Gui/Nameplates/NameplateGui.cs:211]: I feel like there has to be a better way, but we can look into that later, as this is implementation details that we can address at any time without breaking the API.

File: Dalamud/Plugin/Services/INameplatesGui.cs
Comment (MidoriKami) [inline Dalamud/Plugin/Services/INameplatesGui.cs:9]: I would strongly prefer `INameplateGui` as I and many others are likely going to not notice that it's plural, where as other services are singular.

**Review by MidoriKami — CHANGES_REQUESTED** [review]
Some concerns that I think need to be addressed immediately before it can be merged, and some suggestions that would be nice to have.

I won't be available for follow up reviews until I have completed MSQ.

**Comment by nebel** [comment]
> That would be nice, but it'd have to go in this weekend. Could you please re-review if you have time? @Haselnussbomber, @nebel and @MidoriKami - thank you!

There is currently an issue with this service which could be quite difficult to address, namely that the main function we hook to do all the work (`SetPlayerNamePlate`) has been fully inlined in 7.0. The original function still exists but it is no longer called, and our hook does nothing. Until we can figure out how to fix this, I think we need to hold off on merging this.

**Comment by goaaats** [comment]
Thanks for the assessment. I assume this will be a problem for kal too then - maybe a solution can be worked out together.

**Comment by Pilzinsel64** [comment]
Arg, that was unintended... Maybe they moved it to the paren function now. 

Hm, there is another function we could hook to. But I don't know how it's called and not what the sig is. I know it's more low level and is called for every single namplate when updating the stringdata, so also NPCs, if I'm right. But I have absolutely no experience with it.

File: Dalamud.CorePlugin/PluginImpl.cs
Comment (Pilzinsel64) [reply Dalamud.CorePlugin/PluginImpl.cs:1]: Wasn't the plan. I prevented commiting it long time but seems somewhen I just missed it. Reverting it.

File: Dalamud/Game/Gui/Nameplates/Model/INameplateInfo.cs
Comment (Pilzinsel64) [reply Dalamud/Game/Gui/Nameplates/Model/INameplateInfo.cs:37]: Not really. It's readonly. :D

**Comment by Pilzinsel64** [comment]
Deprecated due #1915 (big thanks to SE :/ )

**Pull request closed without merging**
